### ISSUE TITLE
GPO percentile not as expected

### ISSUE THREAD

**Issue by ShibWizard** [ethereum/go-ethereum#30857]
I think gpo percentile is implemented incorrectly here

https://github.com/ethereum/go-ethereum/blob/master/eth/gasprice/gasprice.go#L215

In the code we are calculating percentile of the position rather than percentile of the actual gas, For example if the `results` array contain same gas price like [10,10,10,10], Whatever is the value of percentile it will alway give 10 as the price.

Rather then it should maybe return a percentile of the highest, lower or the middle value from the array ?

**Comment by fishonamos** [comment]
Let me work on this @ShibWizard

**Comment by ShibWizard** [comment]
thanks @fishonamos, I hope my claim makes sense.

**Comment by holiman** [comment]
I disagree. Let's consider the `results` to be a histogram of values. If we ask for the top 1 percentile, in a histogram with `100` values, then we are literally asking "what value would be higher than 99 of these values". 
@ShibWizard could you give a specific case where you think that the current implementation gives an incorrect result, along with what you think it should give? You mentioned the example case of `[10,10,10,10],` but didn't mention what you think would be the proper response (also, `[10,10,10,10]` is an edgecase, because the percentiles on a histogram with no spread is, I would guess, undefined)

**Comment by ShibWizard** [comment]
Yes I agree it's an edgecase, It might not make sense in ethereum where there is a wide spread but for relatively smaller networks with limited no of txns, One can try to manipulate the gas fees from GPO using this edge case.

**Comment by ShibWizard** [comment]
That's why I didn't reported this as an error, I just thought this to be a topic of discussion and wanted to bring to the notice of ethereum developers.

**Comment by ShibWizard** [comment]
One scenario I faced was in a closed private network where one of my team dev (only guy doing txns at the moment) was fetching price from GPO and was always using 20% buffer on top of that, Since he was the only one doing txns the GPO after every block used to return the gas he used as the suggested fees, Although the block were not fully filled and technically he could have just used the min gas configures, He was getting higher and higher suggestion from GPO after every block.

Basically in the results array there is only one response that is [100], So he got 100 as suggestion and used 120, So in the next block now the results have [120] and he used 144 now, This kept happening till 500 Gwei as that's the max and then he reported this to me and I found this edgecase. I know multiple factors are responsible for this, Not alone this part of code.

**Comment by holiman** [comment]
Ok, so you are not saying that the percentiles are wrong, then, but that the edgecase of a no-spread histogram should return a different answer? 

What do you propose that it should return, for the example of `[10, 10, 10, 10]`?

**Comment by holiman** [comment]
> and was always using 20% buffer on top of that

IMO that was the error here, causing the runaway gasprice

**Comment by ShibWizard** [comment]
I agree with you @holiman on the edgecase part.
Regarding what it should return in the case of edgecase, Not sure as of now, I will think about it.
My current motive was to bring it to notice and get advice from you guys, Which I got, So thanks for that.
